src/plugins/vis_builder/public/saved_visualizations/transforms.ts

```diff
@@ -26,3 +30,51 @@ export const saveStateToSavedObject = (
 
   return obj;
 };
+
+export interface VisBuilderSavedVis
+  extends Pick<VisBuilderSavedObjectAttributes, 'id' | 'title' | 'description'> {
+  state: RenderState;
+}
+
+export const getStateFromSavedObject = (
+  obj: VisBuilderSavedObjectAttributes
+): VisBuilderSavedVis => {
+  const { id, title, description } = obj;
+  const styleState = JSON.parse(obj.styleState || '');
+  const vizStateWithoutIndex = JSON.parse(obj.visualizationState || '');
+  const visualizationState: VisualizationState = {
+    searchField: '',
+    ...vizStateWithoutIndex,
+    indexPattern: obj.searchSourceFields?.index,
+  };
+
+  const validateResult = validateVisBuilderState({ styleState, visualizationState });
+
+  if (!validateResult.valid) {
+    throw new InvalidJSONProperty(
+      validateResult.errorMsg ||
+        i18n.translate('visBuilder.getStateFromSavedObject.genericJSONError', {
+          defaultMessage:
+            'Something went wrong while loading your saved object. The object may be corrupted or does not match the latest schema',
+        })
+    );
+  }
+
+  if (!visualizationState.indexPattern) {
+    throw new Error(
+      i18n.translate('visBuilder.getStateFromSavedObject.missingIndexPattern', {
+        defaultMessage: 'The saved object is missing an index pattern',
+      })
+    );
+  }
```

Comment:
Thats going to be a hard state to recover from since the new index pattern should also have similar mappings for the fields, and that goes into the whole rabbit hole of how to recover from a broken saved object state, and thats a much bigger problem to solve for and i wouldnt limit it to just here.